types/index.d.ts

```diff
@@ -101,7 +101,15 @@ declare module 'mongoose' {
     [k: string]: any
   }
 
-  export type Require_id<T> = T extends { _id?: any } ? (T & { _id: T['_id'] }) : (T & { _id: Types.ObjectId });
+  export type Require_id<T> = T extends { _id?: infer U } 
+    ? U extends any
```

Comment:
Sorry that I'm not excellent in Typescript, but what does `U extends any` here mean? 

As far as I know, everything extends any? so the contiditional clause `(T & { _id: Types.ObjectId })` is always used, and the clause `T & Required<{ _id: U }>` is never used. 

Thus, `_id` will be always infered with type `ObjectId`, even if I have defined another type such as `String` in the Schema definition.

Reply:
For example:
```ts
import {model, Schema} from "mongoose";
const schema_with_string_id = new Schema({_id: String, nickname: String})
const theModel = model('test', schema_with_string_id)
const obj = new theModel()
const theId = obj._id
```
The type of variable `theId` will be `string | Types.ObjectId`, rather than `string`. 
Is this designed behaviour for some reason I don't know, or just a bug?

Reply:
@mohammad0-0ahmad Sorry for bothering you!

Reply:
Hello @Starrah !
Firstly, you are not bothering, and you are totally right.
I think this meant to be  something like:
```TS
  export type Require_id<T> = T extends { _id?: infer U }
    ? IfEquals<U, any> extends true
      ? (T & { _id: Types.ObjectId })
      : T & Required<{ _id: U }>
    : T & { _id: Types.ObjectId };
```
Thanks for your review, I will make sure to refactor this.
What are your thoughts @Uzlopak  about that?

Reply:
I think I did this to check if T is not never.

Reply:
> I think I did this to check if T is not never.

First, does you mean you did that to check if `{_id: }` in T is not never, rather than check T?
Because according to my experiment, 
```ts
import {Require_id, Types} from "mongoose";

type Required_id_old<T> = T extends { _id?: any } ? (T & { _id: T['_id'] }) : (T & { _id: Types.ObjectId });
type Required_id_new<T> = T extends { _id?: infer U }
    ? U extends any
        ? (T & { _id: Types.ObjectId })
        : T & Required<{ _id: U }>
    : T & { _id: Types.ObjectId };

type t1_old = Required_id_old<{_id: never}>
type t1_new = Required_id_new<{_id: never}>

type t2_old = Required_id_old<never>
type t2_new = Required_id_new<never>
```
Both `t2_old` and `t2_new` is `never`(which I think is expected behaviour), and for t1, `t1_old` is `{_id: never}` but `t1_new` is `never`(I have no idea which is your expected behaviour).

Then, if my assumption is true and you does mean `check if `{_id: }` in T is not never`, then maybe the two conditional clause should be swapped, as the following:
```ts
type Required_id_new<T> = T extends { _id?: infer U }
    ? U extends any
        ? T & Required<{ _id: U }>
        : (T & { _id: Types.ObjectId })
    : T & { _id: Types.ObjectId };
```

Reply:
@mohammad0-0ahmad @Uzlopak 
Though with the discussion above, we (or just I) cannot still figure out what the codes here means, specifically, what does the `U extends any` mean.  

**However, I want to point out that the following test case written by me, got failed:**
```ts
// index.test-d.ts
import {model, Schema} from "mongoose";
import {expectType} from "tsd";

const schema_with_string_id = new Schema({_id: String, nickname: String})
const theModel = model('test', schema_with_string_id)
const obj = new theModel()

expectType<string>(obj._id)
```
when testing it with `tsd`, gives the following error: 
```
 index.test-d.ts:8:0
  ✖  8:0  Parameter type string is declared too wide for argument type string & ObjectId.
```
According to @mohammad0-0ahmad :


> Hello @Starrah ! Firstly, you are not bothering, and you are totally right. I think this meant to be something like:
> 
> ```ts
>   export type Require_id<T> = T extends { _id?: infer U }
>     ? IfEquals<U, any> extends true
>       ? (T & { _id: Types.ObjectId })
>       : T & Required<{ _id: U }>
>     : T & { _id: Types.ObjectId };
> ```
> 
> Thanks for your review, I will make sure to refactor this. What are your thoughts @Uzlopak about that?

I think that the test case failure above **indicates a bug**. So, I decide to raise a new issue (#12070) and assign @mohammad0-0ahmad, to report and/or further discuss this bug.